mm2src/coins/utxo/utxo_sql_block_header_storage.rs

```diff
@@ -0,0 +1,344 @@
+    let sql = "INSERT OR IGNORE INTO ".to_owned() + &table_name + " (block_height, hex) VALUES (?1, ?2);";
+    Ok(sql)
+}
+
+fn get_block_header_by_height(for_coin: &str) -> Result<String, MmError<BlockHeaderStorageError>> {
+    let table_name = block_headers_cache_table(for_coin);
+    validate_table_name(&table_name).map_err(|e| BlockHeaderStorageError::CantRetrieveTableError {
+        ticker: for_coin.to_string(),
+        reason: e.to_string(),
+    })?;
+
+    let sql = "SELECT hex FROM ".to_owned() + &table_name + " WHERE block_height=?1;";
+
+    Ok(sql)
+}
+
+#[derive(Clone, Debug)]
+pub struct SqliteBlockHeadersStorage(pub Arc<Mutex<Connection>>);
+
+fn query_single_row<T, P, F>(
+    conn: &Connection,
+    query: &str,
+    params: P,
+    map_fn: F,
+) -> Result<Option<T>, MmError<BlockHeaderStorageError>>
+where
+    P: IntoIterator,
+    P::Item: ToSql,
+    F: FnOnce(&Row<'_>) -> Result<T, SqlError>,
+{
+    let maybe_result = conn.query_row(query, params, map_fn);
+    if let Err(SqlError::QueryReturnedNoRows) = maybe_result {
+        return Ok(None);
+    }
+
+    let result = maybe_result.map_err(|e| BlockHeaderStorageError::QueryError {
+        query: query.to_string(),
+        reason: e.to_string(),
+    })?;
+    Ok(Some(result))
+}
+
+fn string_from_row(row: &Row<'_>) -> Result<String, SqlError> { row.get(0) }
+
+#[async_trait]
+impl BlockHeaderStorageOps for SqliteBlockHeadersStorage {
+    async fn init(&self, for_coin: &str) -> Result<(), MmError<BlockHeaderStorageError>> {
+        let selfi = self.clone();
+        let sql_cache = create_block_header_cache_table_sql(for_coin)?;
+        let ticker = for_coin.to_owned();
+        async_blocking(move || {
+            let conn = selfi.0.lock().unwrap();
+            conn.execute(&sql_cache, NO_PARAMS)
+                .map(|_| ())
+                .map_err(|e| BlockHeaderStorageError::ExecutionError {
+                    ticker,
+                    reason: e.to_string(),
+                })?;
+            Ok(())
+        })
+        .await
+    }
+
+    async fn is_initialized_for(&self, for_coin: &str) -> Result<bool, MmError<BlockHeaderStorageError>> {
+        let block_headers_cache_table = block_headers_cache_table(for_coin);
+        validate_table_name(&block_headers_cache_table).map_err(|e| {
+            BlockHeaderStorageError::CantRetrieveTableError {
+                ticker: for_coin.to_string(),
+                reason: e.to_string(),
+            }
+        })?;
+
+        let selfi = self.clone();
+        async_blocking(move || {
+            let conn = selfi.0.lock().unwrap();
+            let cache_initialized = query_single_row(
+                &conn,
+                CHECK_TABLE_EXISTS_SQL,
+                [block_headers_cache_table],
+                string_from_row,
+            )?;
+            Ok(cache_initialized.is_some())
+        })
+        .await
+    }
+
+    async fn add_electrum_block_headers_to_storage(
```

Comment:
It looks quite similar to `add_block_headers_to_storage`. Is it possible to reduce code duplication?